crates/emitter/src/control_structures.rs

```diff
@@ -0,0 +1,360 @@
+            .last_mut()
+            .expect("There should be at least one loop")
+    }
+}
+
+pub struct LoopControl {
+    pub breaks: Vec<BytecodeOffset>,
+    pub continues: Vec<BytecodeOffset>,
+    head: BytecodeOffset,
+}
+
+impl LoopControl {
+    pub fn new(emit: &mut InstructionWriter, depth: u8) -> Self {
+        // Insert a Nop if needed to ensure the script does not start with a
+        // JSOp::LoopHead. This avoids JIT issues with prologue code + try notes
+        // or OSR. See bug 1602390 and bug 1602681.
+        let mut offset = emit.bytecode_offset();
+        if offset.offset == 0 {
+            emit.nop();
+            offset = emit.bytecode_offset();
+        }
+        emit.loop_head(depth);
+        // emit the jump target for the loop head
+        Self {
+            breaks: Vec::new(),
+            continues: Vec::new(),
+            head: offset,
+        }
+    }
+
+    pub fn register_break(&mut self, offset: BytecodeOffset) {
+        // offset points to the location of the jump, which will need to be updated
+        // once we emit the jump target in emit_jump_target_and_patch
+        self.breaks.push(offset);
+    }
+
+    // TODO: fix continues so that they work with scopes correctly
+
+    pub fn emit_end_target(self, emit: &mut InstructionWriter) {
+        let offset = emit.bytecode_offset();
+        // FIXME: we should only need goto_ here
+        emit.goto_(0 as i32);
+        emit.patch_jump_to_target(self.head, offset);
+
+        emit.emit_jump_target_and_patch(self.breaks);
+    }
+}
+
+// Struct for multiple jumps that point to the same target. Examples are breaks and loop conditions.
+pub struct BreakEmitter {
+    pub jump: JumpKind,
+}
+
+impl Jump for BreakEmitter {
+    fn jump_kind(&mut self) -> &JumpKind {
+        &self.jump
+    }
+}
+
+impl BreakEmitter {
+    pub fn emit(&mut self, emitter: &mut AstEmitter) {
+        // break { control, jumpkind }.emit(self)
+        //      -> register break
+        //      -> jump
+        let offset = emitter.emit.bytecode_offset();
+        emitter.loop_stack.register_break(offset);
+        self.emit_jump(emitter);
+    }
+}
+
+pub struct WhileEmitter<F1, F2>
+where
+    F1: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+    F2: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub test: F1,
+    pub block: F2,
+}
+impl<F1, F2> WhileEmitter<F1, F2>
+where
+    F1: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+    F2: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub fn emit(&mut self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
+        // TODO: Set postions
+        // Parameters are the offset in the source code for each
+        // character below:
+        //
+        //   while ( x < 20 ) { ... }
+        //   ^       ^              ^
+        //   |       |              |
+        //   |       |              endPos_
+        //   |       |
+        //   |       condPos_
+        //   |
+        //   whilePos_
```

Comment:
this comment alone doesn't make.
I think it's better just removing this (and others)